File: src/823_Binary_Trees_With_Factors.py
# 823_Binary_Trees_With_Factors
# https://leetcode.cn/problems/binary-trees-with-factors/

from typing import List

class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        m = 10**9 + 7
        arr.sort()
        n = len(arr)
        dp = [1] * n
        for i in range(1, n):
            l = 0
            r = i-1
            tmp = arr[l] * arr[r]
            while l <= r:
                if tmp < arr[i]:
                    l += 1
                elif tmp > arr[i]:
                    r -= 1
                else:
                    if l == r:
                        dp[i] = (dp[i] + (dp[l] * dp[l])) % m
                    else:
                        dp[i] = (dp[i] + 2 * (dp[l] * dp[r])) % m
                    l += 1
                    r -= 1
                tmp = arr[l] * arr[r]

        return sum(dp) % m


# 优化
from typing import List

class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        MOD = 10**9 + 7
        arr.sort()
        
        # 用哈希表记录每个值的索引，以便快速查找
        index = {num: i for i, num in enumerate(arr)}
        
        dp = [1] * len(arr)
        for i, num in enumerate(arr):
            for j in range(i):
                if num % arr[j] == 0: # arr[j]是可能的左子节点
                    right_child = num // arr[j]
                    if right_child in index: # 如果也在arr中，那么它是一个有效的右子节点
                        dp[i] += dp[j] * dp[index[right_child]]
                        dp[i] %= MOD
        
        return sum(dp) % MOD


        
s = Solution()
ans = s.numFactoredBinaryTrees(arr = [2, 4])
print(ans) # 3


s = Solution()
ans = s.numFactoredBinaryTrees(arr = [2, 4, 5, 10])
print(ans) # 7

s = Solution()
ans = s.numFactoredBinaryTrees(arr = [18,3,6,2])
print(ans) # 12

s = Solution()
ans = s.numFactoredBinaryTrees(arr = [45,42,2,18,23,1170,12,41,40,9,47,24,33,28,10,32,29,17,46,11,759,37,6,26,21,49,31,14,19,8,13,7,27,22,3,36,34,38,39,30,43,15,4,16,35,25,20,44,5,48])
print(ans) # 777

File: src/823_Binary_Trees_With_Factors.rs
// 823_Binary_Trees_With_Factors
// https://leetcode.cn/problems/binary-trees-with-factors/

// 给出一个含有不重复整数元素的数组 arr ，每个整数 arr[i] 均大于 1。
// 用这些整数来构建二叉树，每个整数可以使用任意次数。其中：每个非叶结点的值应等于它的两个子结点的值的乘积。
// 满足条件的二叉树一共有多少个？答案可能很大，返回 对 109 + 7 取余 的结果。

// 示例 1:
// 输入: arr = [2, 4]
// 输出: 3
// 解释: 可以得到这些二叉树: [2], [4], [4, 2, 2]

// 示例 2:
// 输入: arr = [2, 4, 5, 10]
// 输出: 7
// 解释: 可以得到这些二叉树: [2], [4], [5], [10], [4, 2, 2], [10, 2, 5], [10, 5, 2].
use std::collections::HashMap;

impl Solution {
    pub fn num_factored_binary_trees(arr: Vec<i32>) -> i64 {
        const MOD: i32 = 1_000_000_007;
        
        let mut arr = arr.clone();
        arr.sort();
        
        let index: HashMap<i32, usize> = arr.iter()
            .enumerate()
            .map(|(i, &num)| (num, i))
            .collect();
        
        let mut dp = vec![1; arr.len()];
        
        for i in 0..arr.len() {
            for j in 0..i {
                if arr[i] % arr[j] == 0 {
                    let right_child = arr[i] / arr[j];
                    if let Some(&right_index) = index.get(&right_child) {
                        dp[i] = (dp[i] + dp[j] as i64 * dp[right_index] as i64 % MOD as i64) % MOD as i64;
                    }
                }
            }
        }
        
        dp.iter().sum::<i64>() % MOD as i64
    }
}